keps/sig-cli/2551-return-code-normalization/README.md

```diff
@@ -0,0 +1,652 @@
+
+```
+<<[UNRESOLVED initial error codes ]>>
+Define the initial error codes, and how to document them
+<<[/UNRESOLVED]>>
+```
+
+## Design Details
+
+The majority of commands already are organized as the following:
+* Run Complete to complete missing information with defaults. This runs on the client side
+* Run Validation to check command syntax and missing arguments. This runs on the client side
+* Run the command itself. This might run on the client side, be dry-run, run an external command or call the APIServer.
+
+One thing that should be done is map all the error codes/ints as constants in some file, so they can be automatically
+documented.
+
+There are three design suggestions:
+
+```
+<<[UNRESOLVED select design approach ]>>
+Select the best design approach, move others to design alternatives
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED dealing with kubectl exec and run exit ]>>
+kubectl exec and run uses the pod exit code as its own exit code, we should figure out how we 
+should deal with this
+https://github.com/kubernetes/kubernetes/blob/v1.22.0-alpha.0/test/e2e/kubectl/kubectl.go#L496
+https://github.com/kubernetes/kubernetes/blob/v1.22.0-alpha.0/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L178-L179
+<<[/UNRESOLVED]>>
+```
+
+### Changing error checker functions
+All of the steps above uses [cmdutil.CheckErr](https://github.com/kubernetes/kubernetes/blob/2a26f276a8c8c13b2f45927ee5ece2063950dd1d/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L114) 
+function to delegate the error validation. This function might be slight changed 
+(without changing its signature) to instead of use the default `fatalErrHandler` 
+function, use some more specific function that might verify the error against a matrix of possible errors
+and exit with the right code.
+
+The function [CheckDiffErr](https://github.com/kubernetes/kubernetes/blob/2a26f276a8c8c13b2f45927ee5ece2063950dd1d/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L124)
+can be used as an example of how this can be implemented.
+
+The pro of using this design approach is that almost no change will be needed in commands, as they already use the 
+CheckErr function.
+
+The con of using this design approach is that the Error Validation functions are pretty hard to change/evolve
+in the manner they are designed, calling another checkErr inner function.
+
+For the rollout of this approach, a new non-exported error checker function needs to be developed and make the 
+call to the old checkErr or the new function according to the value of the Environment Variable described in
+[Risks and Mitigations](#risks-and-mitigations) 
+
+### Creating new error parser functions
+
+Another design solution is to create helper functions for each steps:
+* When running Complete, Validate or other client side steps, call cmdutil.CheckClientErr(err) and exits with some well defined client error code, mapped to ErrorCodeClient
```

Comment:
As mentioned before I'd describe this in slightly more generic way, since this implementation might change.

Reply:
@deejross to take a look into this